tests/components/smartthings/test_config_flow.py

```diff
@@ -8,434 +8,720 @@
+        result = await hass.config_entries.flow.async_configure(result["flow_id"])
+        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+        assert result["data"]["app_id"] == app.app_id
+        assert result["data"]["installed_app_id"] == installed_app_id
+        assert result["data"]["location_id"] == location.location_id
+        assert result["data"]["access_token"] == token
+        assert result["data"]["refresh_token"] == request.refresh_token
+        assert result["data"]["client_secret"] == app_oauth_client.client_secret
+        assert result["data"]["client_id"] == app_oauth_client.client_id
+        assert result["title"] == location.name
+        entry = next(
+            (entry for entry in hass.config_entries.async_entries(DOMAIN)), None,
+        )
+        assert entry.unique_id == smartapp.format_unique_id(
+            app.app_id, location.location_id
+        )
+
+
+async def test_invalid_webhook_aborts(hass):
+    """Test flow aborts if webhook is invalid."""
+    hass.config.api.base_url = "http://0.0.0.0"
+
+    # Webhook confirmation shown
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "user"}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "invalid_webhook_url"
+    assert result["description_placeholders"][
+        "webhook_url"
+    ] == smartapp.get_webhook_url(hass)
     assert "component_url" in result["description_placeholders"]
 
 
-async def test_step_pat_invalid_token(hass):
+async def test_invalid_token_shows_error(hass):
     """Test an error is shown for invalid token formats."""
-    flow = SmartThingsFlowHandler()
-    flow.hass = hass
     token = "123456789"
 
-    result = await flow.async_step_pat({CONF_ACCESS_TOKEN: token})
+    # Webhook confirmation shown
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "user"}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    assert result["description_placeholders"][
+        "webhook_url"
+    ] == smartapp.get_webhook_url(hass)
+
+    # Advance to PAT screen
+    result = await hass.config_entries.flow.async_configure(result["flow_id"], {})
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "pat"
+    assert "token_url" in result["description_placeholders"]
+    assert "component_url" in result["description_placeholders"]
 
+    # Enter token
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {CONF_ACCESS_TOKEN: token}
+    )
     assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
     assert result["step_id"] == "pat"
     assert result["data_schema"]({}) == {CONF_ACCESS_TOKEN: token}
-    assert result["errors"] == {"access_token": "token_invalid_format"}
+    assert result["errors"] == {CONF_ACCESS_TOKEN: "token_invalid_format"}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5444210)